core/src/main/java/com/datastax/oss/driver/internal/core/cql/CqlRequestHandlerBase.java

```diff
@@ -216,13 +216,22 @@ public void onThrottleReady(boolean wasDelayed) {
 
   private Timeout scheduleTimeout(Duration timeoutDuration) {
     if (timeoutDuration.toNanos() > 0) {
-      return this.timer.newTimeout(
-          (Timeout timeout1) -> {
-            setFinalError(
-                new DriverTimeoutException("Query timed out after " + timeoutDuration), null, -1);
-          },
-          timeoutDuration.toNanos(),
-          TimeUnit.NANOSECONDS);
+      try {
+        return this.timer.newTimeout(
+            (Timeout timeout1) ->
+                setFinalError(
+                    new DriverTimeoutException("Query timed out after " + timeoutDuration),
+                    null,
+                    -1),
+            timeoutDuration.toNanos(),
+            TimeUnit.NANOSECONDS);
+      } catch (IllegalStateException e) {
+        // If we raced with session shutdown the timer might be closed already, rethrow with a more
+        // explicit message
+        throw ("cannot be started once stopped".equals(e.getMessage()))
+            ? new IllegalStateException("Session is closed")
+            : e;
```

Comment:
Unfortunately there's not better way to detect that the timer was stopped.

We didn't run into this in driver 3, because timeouts were scheduled per node, and always after the initial write.

Reply:
I added `ShutdownIT`, it fails if I replace this line with `throw e`.